enhancements/cluster-high-availability-mode-api.md

```diff
@@ -0,0 +1,401 @@
+---
+title: cluster-high-availability-mode-api
+authors:
+  - "@dhellmann"
+  - "@eranco"
+  - "@mrunalp"
+  - "@romfreiman"
+reviewers:
+  - TBD
+  - "@markmc"
+  - "@deads2k"
+  - "@wking"
+  - "@hexfusion"
+approvers:
+  - TBD
+  - "@eparis"
+  - "@derekwaynecarr"
+creation-date: 2020-12-02
+last-updated: 2020-12-02
+status: implementable
+see-also:
+  - "/enhancements/single-node-production-edge-cluster-profile.md"
+  - "/enhancements/update/cluster-profiles.md"
+---
+
+# Cluster High-availability Mode API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As we add more deployment topologies for OpenShift, we need a way to
+communicate expectations to the operators managing the cluster. This
+enhancement describes an API for telling operators how hard to work to
+provide high-availability of their operands.
+
+## Motivation
+
+We want to avoid having cluster profiles use different manifests to
+configure operators managed by the `cluster-version-operator`, and
+instead rely on in-cluster APIs. This approach will make the same
+information available to operators managed by either the
+`cluster-version-operator` or `operator-lifecycle-manager`.
+
+We do not want to expose the cluster profile name directly to most of
+the operators, because we do not want to have to update all of the
+operators each time we add a new profile.
+
+We want the setting to be global to the cluster, rather than using
+operator-specific APIs, because that simplifies the integration with
+the installer and allows it to be applied consistently in all
+operators that need them.
+
+### Goals
+
+- Expose high-availability expectations to operators running in a
+  cluster so they can configure their operands appropriately.
+
+### Non-Goals
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the high-availability mode (at
+  least for now).
+- Predict other capability settings that we might need to expose.
+- Define a more generic "capabilities" API.
+
+## Proposal
+
+### Cluster Operators
+
+The `infrastructure.config.openshift.io` [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+is expanded to include two new fields to describe the
+high-availability mode desired for cluster operators running on the
+cluster through "topology" values for the control plane and
+infrastructure, separately.
+
+The `controlPlaneTopology` field will express the expectations for
```

Comment:
how does the installer expected to set those values:
masters=1,workers=0: controlPlaneTopology=SingleNode,infrastructureTopology=None?
masters=1,workers=1: controlPlaneTopology=SingleNode,infrastructureTopology=SingleNode
masters=1,workers>=2: controlPlaneTopology=SingleNode,infrastructureTopology=HighlyAvailable

and if the default is HighlyAvailable, how will we upgrade master=3,worker=1 cluster? (if there are such)

Reply:
Today a cluster with 3 control plane nodes and 1 worker nodes is not supported in the installer. It fails to deploy because ingress can't create enough replicas of its operand.

Reply:
I've added some prose explaining the rules for setting the 2 topology fields on lines 97-107 of the latest draft. Please review.